src/contrast.rs

```diff
@@ -7,9 +8,86 @@ use image::{
     Luma
 };
 
-/// Returns the cumulative histogram of grayscale values in an 8bpp
+/// Returns the Otsu threshold level of an 8bpp image.
+/// This threshold will optimally binarize an image that
+/// meets certain criteria.
+pub fn otsu_level<I>(image: &I) -> u8
+    where I: GenericImage<Pixel=Luma<u8>> {
+    let hist = histogram(image);
+    let levels = hist.len();
+    let mut histc = [0i32; 256];
+    let mut meanc = [0.0f64; 256];
+    let mut pixel_count = hist[0] as f64;
+
+
+    histc[0] = hist[0];
+
+    for i in 1..levels {
+        pixel_count += hist[i] as f64;
+        histc[i] = histc[i-1] + hist[i];
+        meanc[i] = meanc[i-1] + (hist[i] as f64)*(i as f64);
+    }
+
+    let mut sigma_max:f64 = -1.0;
+    let mut otsu:f64 = 0.0;
+    let mut otsu2:f64 = 0.0;
+
+    for i in 0..levels {
+        meanc[i] /= pixel_count;
+
+        let p0 = (histc[i] as f64)/pixel_count;
+        let p1 = 1.0f64 - p0;
+        let mu0 = meanc[i]/p0;
+        let mu1 = (meanc[levels-1]/pixel_count - meanc[i])/p1;
+
+        let sigma = p0*p1*(mu0-mu1).powi(2);
+        if sigma >= sigma_max {
+            if sigma > sigma_max {
+                otsu = i as f64;
+                sigma_max = sigma;
+            } else {
+                otsu2 = i as f64;
+            }
+        }
+    }
+    ( (otsu + otsu2) / 2.0 ).ceil() as u8
+}
+
+/// Returns a binarized image from an input 8bpp grayscale image
+/// obtained by applying the given image
+pub fn threshold<I>(image: &I, thresh: u8) -> GrayImage
+    where I: GenericImage<Pixel=Luma<u8>> {
+    let mut out: GrayImage = ImageBuffer::new(image.width(), image.height());
+     for y in 0..image.height() {
+            for x in 0..image.width() {
+                if image.get_pixel(x, y)[0] as u8 <= thresh {
+                    out.put_pixel(x, y, Luma([0 as u8]));
+                } else {
+                    out.put_pixel(x, y, Luma([255 as u8]));
+                }
+            }
+        }
+    out
+}
+
+/// Mutates given image to form a binarized version produced by applying
+/// the given threshold
+pub fn threshold_mut<I>(image: &mut I, thresh: u8)
+    where I: GenericImage<Pixel=Luma<u8>>  {
+     for y in 0..image.height() {
```

Comment:
When there's a pair of f and f_mut, I'd prefer that f_mut contain the full logic, and the body of f just be: "let out = ImageBuffer::new(...); out.copy_from(input, 0, 0); f(out); out". See e.g equalize_histogram and equalize_histogram_mut.